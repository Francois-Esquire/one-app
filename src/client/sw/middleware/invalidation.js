/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import {
  isCacheStorageSupported,
  isServiceWorker,
  getMetaData,
  setMetaData,
  remove,
} from '@americanexpress/one-service-worker';

import {
  getResourceTypeFromCacheName, getModuleInfoFromUrl, getLangPackInfoFromUrl, markForRemoval,
} from './utility';

export default function createInvalidationMiddleware() {
  if (!isCacheStorageSupported() || !isServiceWorker()) return function noop() {};

  return function invalidationMiddleware(event, context) {
    const { request, cacheName } = context.get();

    if (request && request instanceof Request) {
      const { url } = request;

      const type = getResourceTypeFromCacheName(cacheName);

      let info = {};

      switch (type) {
        case 'one-app':
          info = getLangPackInfoFromUrl(url);
          break;
        case 'module':
          info = getModuleInfoFromUrl(url);
          break;
        case 'language-pack':
          info = getLangPackInfoFromUrl(url);
          break;
        default:
          break;
      }

      const resourceKey = new Request(`/${type}/${info.name}`).url;
      const resourceMeta = {
        ...info,
        type,
        url,
      };

      event.waitUntil(
        getMetaData({
          url: resourceKey,
        }).then((meta) => {
          if (markForRemoval(meta, resourceMeta)) {
            event.waitUntil(remove(new Request(meta.url)));
          }

          return setMetaData({
            url: resourceKey,
            metadata: resourceMeta,
          });
        })
      );
    }
  };
}
