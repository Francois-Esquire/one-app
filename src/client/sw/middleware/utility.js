/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export function getConfig() {
  return JSON.parse(process.env.OSW_CONFIG || '{}');
}

export function getResourceTypeFromCacheName(cacheName = '') {
  const [, type = ''] = cacheName.match(/.*(one-app|language-pack|module)-cache$/) || [];
  return type;
}

export function getClientCacheRevisionFromUrl(url = '') {
  const [, key = ''] = url.match(/^https?.*(?:\.js\?clientCacheRevision=(.*))$/) || [];
  return key;
}

export function getLangPackInfoFromUrl(url = '') {
  // capture groups: name, version, locale, language, country, env,
  const info = url.match(/^https?.*\/(?<name>.*)\/(?<version>.*)\/(?<locale>(?<language>[a-z]{2,3})-(?<country>[A-Z]{1,4})?)\/(?<env>qa|integration|.*)\.json$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function getModuleInfoFromUrl(url = '') {
  // capture groups: checksum, name, version, bundle, legacy, revision
  const info = url.match(/^https?.*\/(?<checksum>.*)\/(?<name>.*)\/(?<version>.*)\/(?:.*)\2\.(?<bundle>(legacy\.)?browser)\.js(?:\?clientCacheRevision=(?<revision>.*))?$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function getAppInfoFromUrl(url = '') {
  // capture groups: version, name, i18n
  const info = url.match(/^https?.*\/_\/static\/app\/(?<version>.*)\/(?<name>(?<i18n>i18n)?.*)\.js$/);

  if (info) {
    return info.groups;
  }

  return {};
}

export function markForRemoval(cachedMetaRecord, newMetaRecord) {
  if (cachedMetaRecord.revision !== newMetaRecord.revision) return true;
  if (cachedMetaRecord.type === newMetaRecord.type) {
    if (cachedMetaRecord.name === newMetaRecord.name) {
      if (cachedMetaRecord.version !== newMetaRecord.version) return true;
      if (cachedMetaRecord.locale !== newMetaRecord.locale) return false;
    }
  }
  return false;
}
