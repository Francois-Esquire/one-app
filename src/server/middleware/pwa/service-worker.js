/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import fs from 'fs';

let serviceWorkerScript = null;
let serviceWorkerScope = null;
let serviceWorkerType = null;
let serviceWorkerEnabled = false;

export function getServiceWorkerEnabled() {
  return serviceWorkerEnabled;
}

export function getServiceWorkerScope() {
  return serviceWorkerScope;
}

export function getServiceWorkerType() {
  return serviceWorkerType;
}

export function getServiceWorkerScript() {
  return serviceWorkerScript;
}

export function setServiceWorkerScript(value, scope) {
  if (scope) serviceWorkerScope = scope;
  else if (value) serviceWorkerScope = '/';
  else serviceWorkerScope = null;
  serviceWorkerScript = value;
  return serviceWorkerScript;
}

export function createServiceWorkerNoopScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.noop.js'].join('/')).toString();
}

export function createServiceWorkerScript() {
  // this file is created during build inside lib/server/middleware/pwa
  return fs.readFileSync([__dirname, 'scripts/sw.js'].join('/')).toString();
}

export function createServiceWorkerEscapeHatchScript() {
  return 'self.unregister();';
}

export function configureServiceWorker({
  enabled, escapeHatch, noop, scope,
} = {}) {
  if (escapeHatch) {
    serviceWorkerType = 'escape-hatch';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerEscapeHatchScript(), scope);
  } else if (noop) {
    serviceWorkerType = 'noop';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerNoopScript(), scope);
  } else if (enabled) {
    serviceWorkerType = 'standard';
    serviceWorkerEnabled = true;
    setServiceWorkerScript(createServiceWorkerScript(), scope);
  } else {
    serviceWorkerType = null;
    serviceWorkerEnabled = false;
    setServiceWorkerScript(null);
  }
}

export function serviceWorkerMiddleware() {
  return function serviceWorkerMiddlewareHandler(req, res, next) {
    if (serviceWorkerEnabled === false) return next();
    return res
      .type('js')
      .set('Service-Worker-Allowed', serviceWorkerScope)
      .set('Cache-Control', 'no-store, no-cache')
      .send(getServiceWorkerScript());
  };
}
