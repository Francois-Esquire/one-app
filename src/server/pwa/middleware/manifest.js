/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let webmanifest = null;
let webmanifestEnabled = false;

export function resetWebManifest() {
  webmanifest = {
    name: 'One App',
    short_name: 'One App',
  };
  return webmanifest;
}

export function getWebManifestEnabled() {
  return webmanifestEnabled;
}

export function getWebManifest() {
  return webmanifest;
}

export function setWebManifest(manifest) {
  if (manifest === null) webmanifest = null;
  else Object.assign(resetWebManifest(), manifest);
  return getWebManifest();
}

export function configureWebManifest({ enabled = false, manifest } = {}) {
  webmanifestEnabled = enabled;
  if (enabled === false) setWebManifest(null);
  else setWebManifest(manifest);
}

export function webmanifestMiddleware() {
  return function webmanifestMiddlewareHandler(req, res, next) {
    if (!webmanifestEnabled) return next();
    return res
      .type('application/manifest+json')
      .send(getWebManifest());
  };
}
