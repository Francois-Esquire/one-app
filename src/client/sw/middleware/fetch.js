/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { expiration, cacheRouter } from '@americanexpress/one-service-worker';

import createInvalidationMiddleware from './invalidation';
import { getConfig } from './utility';

export default function createFetchMiddleware(config = getConfig()) {
  return [
    {
      cacheName: 'one-app-cache',
      // eslint-disable-next-line no-useless-escape
      match: new RegExp(`^https?.*\/_\/static\/app\/${config.buildVersion}\/.*\\.js$`),
    },
    {
      cacheName: 'module-cache',
      // ends with .browser.js and matches the name
      match: /^https?.*\.(.*)\/(.*)\/(?:.*)\1(?:legacy\.)?browser\.js(?:\?clientCacheRevision=.*)?$/,
    },
    {
      cacheName: 'language-pack-cache',
      // any valid bcp 47 locale id
      match: /^https?.*\/.{2,3}-.{1,4}\/(qa|integration|.*)\.json$/,
    },
  ].map(cacheRouter).concat([
    createInvalidationMiddleware(),
    expiration(),
  ]);
}
