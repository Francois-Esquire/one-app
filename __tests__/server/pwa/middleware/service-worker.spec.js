/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import * as serviceWorker from '../../../../src/server/pwa/middleware/service-worker';

const {
  serviceWorkerMiddleware,
  getServiceWorkerEnabled,
  getServiceWorkerScript,
  getServiceWorkerScope,
  getServiceWorkerType,
  setServiceWorkerScript,
  configureServiceWorker,
  createServiceWorkerScript,
  createServiceWorkerEscapeHatchScript,
  createServiceWorkerNoopScript,
} = serviceWorker;

jest.mock('fs', () => ({
  readFileSync: (filePath) => ({ toString: () => (filePath.endsWith('noop.js') ? '[service-worker-noop-script]' : '[service-worker-script]') }),
}));

describe('service worker middleware components', () => {
  test('should consistently export', () => {
    expect(Object.entries(serviceWorker)).toMatchSnapshot();
  });

  test('gets default values', () => {
    expect(getServiceWorkerEnabled()).toBe(false);
    expect(getServiceWorkerScript()).toBe(null);
    expect(getServiceWorkerScope()).toBe(null);
    expect(getServiceWorkerType()).toBe(null);
  });

  test('sets the service worker script and scope', () => {
    const script = '[service-worker-script]';
    const scope = '/my-scope';
    expect(setServiceWorkerScript(script, scope)).toEqual(script);
    expect(getServiceWorkerScope()).toEqual(scope);
    expect(getServiceWorkerScript()).toEqual(script);
  });

  test('sets the service worker script without scope', () => {
    const script = '[service-worker-script-without-scope-set]';
    const scope = '/';
    expect(setServiceWorkerScript(script)).toEqual(script);
    expect(getServiceWorkerScope()).toEqual(scope);
    expect(getServiceWorkerScript()).toEqual(script);
  });

  test('sets the service worker script to null', () => {
    const script = null;
    const scope = null;
    expect(setServiceWorkerScript(script)).toEqual(script);
    expect(getServiceWorkerScope()).toEqual(scope);
    expect(getServiceWorkerScript()).toEqual(script);
  });

  test('creates various types of service worker scripts', () => {
    expect(createServiceWorkerNoopScript()).toEqual('[service-worker-noop-script]');
    expect(createServiceWorkerScript()).toEqual('[service-worker-script]');
    expect(createServiceWorkerEscapeHatchScript()).toEqual('self.unregister();');
  });

  test('configuring the service worker without parameters will disable it', () => {
    expect(configureServiceWorker()).toBeUndefined();
    expect(getServiceWorkerScript()).toBe(null);
    expect(getServiceWorkerScope()).toBe(null);
    expect(getServiceWorkerEnabled()).toBe(false);
  });

  describe('middleware and various configurations', () => {
    beforeAll(configureServiceWorker);

    test('middleware factory returns function', () => {
      expect(serviceWorkerMiddleware()).toBeInstanceOf(Function);
    });

    test('middleware calls next when disabled', () => {
      const middleware = serviceWorkerMiddleware();
      const next = jest.fn();
      expect(middleware(null, null, next)).toBeUndefined();
      expect(next).toHaveBeenCalledTimes(1);
    });

    test('middleware responds with service worker script', () => {
      const middleware = serviceWorkerMiddleware();
      const next = jest.fn();
      const res = {};
      res.send = jest.fn(() => res);
      res.set = jest.fn(() => res);
      res.type = jest.fn(() => res);
      configureServiceWorker({ enabled: true });

      expect(middleware(null, res, next)).toBe(res);

      expect(res.send).toHaveBeenCalledTimes(1);
      expect(res.type).toHaveBeenCalledTimes(1);
      expect(res.set).toHaveBeenCalledTimes(2);
      expect(next).not.toHaveBeenCalled();
      expect(res.type).toHaveBeenCalledWith('js');
      expect(res.set).toHaveBeenCalledWith('Service-Worker-Allowed', getServiceWorkerScope());
      expect(res.set).toHaveBeenCalledWith('Cache-Control', 'no-store, no-cache');
      expect(res.send).toHaveBeenCalledWith(getServiceWorkerScript());
    });

    test('middleware responds with service worker noop script', () => {
      const middleware = serviceWorkerMiddleware();
      const next = jest.fn();
      const res = {};
      res.send = jest.fn(() => res);
      res.set = jest.fn(() => res);
      res.type = jest.fn(() => res);
      configureServiceWorker({ noop: true });

      expect(middleware(null, res, next)).toBe(res);

      expect(res.send).toHaveBeenCalledTimes(1);
      expect(res.type).toHaveBeenCalledTimes(1);
      expect(res.set).toHaveBeenCalledTimes(2);
      expect(next).not.toHaveBeenCalled();
      expect(res.type).toHaveBeenCalledWith('js');
      expect(res.set).toHaveBeenCalledWith('Service-Worker-Allowed', getServiceWorkerScope());
      expect(res.set).toHaveBeenCalledWith('Cache-Control', 'no-store, no-cache');
      expect(res.send).toHaveBeenCalledWith(createServiceWorkerNoopScript());
    });

    test('middleware responds with service worker escape hatch script', () => {
      const middleware = serviceWorkerMiddleware();
      const next = jest.fn();
      const res = {};
      res.send = jest.fn(() => res);
      res.set = jest.fn(() => res);
      res.type = jest.fn(() => res);
      configureServiceWorker({ escapeHatch: true });

      expect(middleware(null, res, next)).toBe(res);

      expect(res.send).toHaveBeenCalledTimes(1);
      expect(res.type).toHaveBeenCalledTimes(1);
      expect(res.set).toHaveBeenCalledTimes(2);
      expect(next).not.toHaveBeenCalled();
      expect(res.type).toHaveBeenCalledWith('js');
      expect(res.set).toHaveBeenCalledWith('Service-Worker-Allowed', getServiceWorkerScope());
      expect(res.set).toHaveBeenCalledWith('Cache-Control', 'no-store, no-cache');
      expect(res.send).toHaveBeenCalledWith(createServiceWorkerEscapeHatchScript());
    });
  });
});
