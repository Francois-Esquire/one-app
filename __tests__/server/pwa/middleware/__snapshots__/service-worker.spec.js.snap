// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service worker middleware components should consistently export 1`] = `
Array [
  Array [
    "getServiceWorkerEnabled",
    [Function],
  ],
  Array [
    "getServiceWorkerScope",
    [Function],
  ],
  Array [
    "getServiceWorkerType",
    [Function],
  ],
  Array [
    "getServiceWorkerScript",
    [Function],
  ],
  Array [
    "setServiceWorkerScript",
    [Function],
  ],
  Array [
    "createServiceWorkerNoopScript",
    [Function],
  ],
  Array [
    "createServiceWorkerScript",
    [Function],
  ],
  Array [
    "createServiceWorkerEscapeHatchScript",
    [Function],
  ],
  Array [
    "configureServiceWorker",
    [Function],
  ],
  Array [
    "serviceWorkerMiddleware",
    [Function],
  ],
]
`;
